File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.java
class Solution {
    public Node connect(Node root) {
        if (root == null) {
            return root;
        }

        Queue<Node> que = new ArrayDeque<>();
        que.add(root);

        while(!que.isEmpty()) {
            int n = que.size();
            Node before = null;

            for (int i = 0; i < n; i++) {
                Node now = que.poll();
                if (before != null) {
                    now.next = before;
                }
                before = now;
                if (now.right != null) {
                    que.add(now.right);
                }
                if (now.left != null) {
                    que.add(now.left);
                }
            }
        }

        return root;
    }
}

File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.py
from collections import deque


class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root is None: return root

        que = deque()
        que.append(root)

        while que:
            n = len(que)
            before = None
            for i in range(n):
                now = que.popleft()
                if before: now.next = before
                before = now
                if now.right: que.append(now.right)
                if now.left: que.append(now.left)

        return root